packages/router/history.ts

```diff
@@ -588,10 +597,43 @@ function getUrlBasedHistory(
   let action = Action.Pop;
   let listener: Listener | null = null;
 
+  let index = getIndex()!;
+  // Index should only be null when we initialize. If not, it's because the
+  // user called history.pushState or history.replaceState directly, in which
+  // case we should log a warning as it will result in bugs.
+  if (index == null) {
+    index = 0;
+    globalHistory.replaceState({ ...globalHistory.state, idx: index }, "");
```

Comment:
This `replaceState()` line is causing a new regression for my project when testing with Jest and JSDom. I observed the following series of circumstances:

1. The test runs in React strict mode, causing the React Router to mount twice.

2. When the React Router is loaded the 1st time, `window.location.href` is `http://www.example.com/path/to/route`

3. When the React Router is loaded the 1st time, `index` is `null`.

4. The `replaceState` changes `window.location.href` to `http://localhost/`. Notice both the origin and the path has changed.

5. When the React Router is loaded the 2nd time (strict mode), the URL is now wrong, causing the wrong route to load and my test to fail.

Reply:
I propose the following which fixes my problem: https://github.com/remix-run/react-router/pull/10001